// @rx-angular/template/esm2022/experimental/virtual-scrolling/lib/virtual-scroll-element.directive.mjs
import { Directive, ElementRef, inject } from '@angular/core';
import { RxVirtualScrollElement } from './model';
import { unpatchedScroll } from './util';
import * as i0 from "@angular/core";
export class RxVirtualScrollElementDirective {
    elementRef = inject(ElementRef);
    elementScrolled$ = unpatchedScroll(this.elementRef.nativeElement);
    getElementRef() {
        return this.elementRef;
    }
    measureOffset() {
        return this.elementRef.nativeElement.getBoundingClientRect().top;
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: RxVirtualScrollElementDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.1", type: RxVirtualScrollElementDirective, isStandalone: true, selector: "[rxVirtualScrollElement]", host: { classAttribute: "rx-virtual-scroll-element" }, providers: [
            {
                provide: RxVirtualScrollElement,
                useExisting: RxVirtualScrollElementDirective,
            },
        ], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: RxVirtualScrollElementDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[rxVirtualScrollElement]',
                    providers: [
                        {
                            provide: RxVirtualScrollElement,
                            useExisting: RxVirtualScrollElementDirective,
                        },
                    ],
                    // eslint-disable-next-line @angular-eslint/no-host-metadata-property
                    host: {
                        class: 'rx-virtual-scroll-element',
                    },
                    standalone: true,
                }]
        }] });
