// @rx-angular/template/esm2022/experimental/virtual-scrolling/lib/virtual-scroll-window.directive.mjs
import { DOCUMENT } from '@angular/common';
import { Directive, ElementRef, inject } from '@angular/core';
import { RxVirtualScrollElement } from './model';
import { unpatchedScroll } from './util';
import * as i0 from "@angular/core";
export class RxVirtualScrollWindowDirective {
    document = inject(DOCUMENT);
    elementRef = new ElementRef(this.document.documentElement);
    elementScrolled$ = unpatchedScroll(this.document);
    getElementRef() {
        return this.elementRef;
    }
    measureOffset() {
        return 0;
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: RxVirtualScrollWindowDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.0.1", type: RxVirtualScrollWindowDirective, isStandalone: true, selector: "rx-virtual-scroll-viewport[scrollWindow]", providers: [
            {
                provide: RxVirtualScrollElement,
                useExisting: RxVirtualScrollWindowDirective,
            },
        ], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.0.1", ngImport: i0, type: RxVirtualScrollWindowDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: 'rx-virtual-scroll-viewport[scrollWindow]',
                    providers: [
                        {
                            provide: RxVirtualScrollElement,
                            useExisting: RxVirtualScrollWindowDirective,
                        },
                    ],
                    standalone: true,
                }]
        }] });
